packages/svelte/src/chat.svelte.ts

```diff
@@ -56,41 +72,65 @@ export class Chat<MESSAGE_METADATA = unknown> {
    * - `ready`: The full response has been received and processed; a new user message can be submitted.
    * - `error`: An error occurred during the API request, preventing successful completion.
    */
-  get status() {
-    return this.#store.status;
+  get status(): ChatStatus {
+    this.#subscribe();
+    return this.#chatStore.getStatus(this.chatId);
   }
-
-  /** The error object of the API request */
-  get error() {
-    return this.#store.error;
+  set status(value: ChatStatus) {
+    this.#chatStore.setStatus({ id: this.chatId, status: value });
   }
 
-  /** The current value of the input. Writable, so it can be bound to form inputs. */
-  input = $state<string>()!;
-
-  /**
-   * Current messages in the chat.
-   *
-   * This is writable, which is useful when you want to edit the messages on the client, and then
-   * trigger {@link reload} to regenerate the AI response.
-   */
-  get messages(): UIMessage<MESSAGE_METADATA>[] {
-    return this.#store.messages;
+  /** The error object of the API request */
+  get error(): Error | undefined {
+    this.#subscribe();
+    return this.#chatStore.getError(this.chatId);
   }
-  set messages(value: UIMessage<MESSAGE_METADATA>[]) {
-    untrack(() => (this.#store.messages = value));
+  set error(value: Error | undefined) {
+    this.#chatStore.setStatus({
+      id: this.chatId,
+      status: 'error',
+      error: value,
+    });
   }
 
-  constructor(options: ChatOptions<MESSAGE_METADATA> = {}) {
-    if (hasChatContext()) {
-      this.#keyedStore = getChatContext() as KeyedChatStore<MESSAGE_METADATA>;
+  constructor(
+    options: () => ChatOptions<
+      MESSAGE_METADATA,
+      UI_DATA_PART_SCHEMAS
+    > = () => ({}),
+  ) {
+    this.#options = $derived.by(options);
+    this.#generateId = $derived(this.#options.generateId ?? generateId);
+    this.chatId = $derived(this.#options.chatId ?? this.#generateId());
+
+    if (this.#options.chatStore) {
+      this.#chatStore = this.#options.chatStore;
+    } else if (hasChatStoreContext()) {
+      this.#chatStore = getChatStoreContext() as ChatStore<
+        MESSAGE_METADATA,
+        UI_DATA_PART_SCHEMAS
+      >;
     } else {
-      this.#keyedStore = new KeyedChatStore<MESSAGE_METADATA>();
+      this.#chatStore = defaultChatStore<
+        MESSAGE_METADATA,
+        UI_DATA_PART_SCHEMAS
+      >({
+        api: '/api/chat',
+        generateId: this.#options.generateId || generateId,
+      });
     }
 
-    this.#options = options;
-    this.messages = options.initialMessages ?? [];
-    this.input = options.initialInput ?? '';
+    this.input = this.#options.initialInput ?? '';
+
+    this.#subscribe = createSubscriber(update => {
+      return this.#chatStore.subscribe({
+        onChatChanged: update,
+      });
+    });
+
+    if (!this.#chatStore.hasChat(this.chatId)) {
+      this.#chatStore.addChat(this.chatId, this.messages);
```

Comment:
@elliott-with-the-longest-name-on-github this was the resolution for the `id` weirdness we were seeing in the tests! Essentially, generate id for the `this.#chatId` (which is wrapped in the `$derived` earlier in the constructor) was getting called after the initial message was constructed so we'd get:

- first message id = 0
- chat id = 1